add CONFIG_CGROUP_BPF to required kernel config

pacoxu opened the pull request:
References:
- https://github.com/containerd/containerd/pull/3799#issuecomment-555740694
- https://man7.org/linux/man-pages/man7/cgroups.7.html
- https://www.kernel.org/doc/Documentation/admin-guide/cgroup-v2.rst
- https://www.kernel.org/doc/Documentation/admin-guide/cgroup-v1/

Fixes #36

/cc @KentaTada @neolit123

pacoxu commented:
As cgroup v1 is in maintenance mode but not removed, we should not remove the cgroup v1 related kernel config out.

However, should we show different message for `CONFIG_CGROUP_BPF` missing?
Or should we check if the user is in cgroup v1 or v2 at this point to decide which check policy to use?

neolit123 commented on validators/types_unix.go:
```diff
@@ -40,11 +40,12 @@ var DefaultSysSpec = SysSpec{
 			{Name: "UTS_NS"},
 			{Name: "CGROUPS"},
-			{Name: "CGROUP_CPUACCT"},
+			{Name: "CGROUP_BPF"},     // cgroups v2
```
do you know where does it say that v2 requires CGROUP_BPF?
no info here:
https://docs.kernel.org/admin-guide/cgroup-v2.html

pacoxu replied:
`device` controller is based on BPF. Let me check if there is more official information about that.

pacoxu replied:
```
Cgroup v2 device controller has no interface files and is implemented
on top of cgroup BPF. 
```
From https://www.kernel.org/doc/Documentation/cgroup-v2.txt.

neolit123 reviewed (commented):
in https://github.com/kubernetes/system-validators/issues/36 @KentaTada mentioned that we need to read another file to verify cgroupv2.

> Unlike in v1, the kernel config alone cannot determine whether the required v2 features are enabled.

pacoxu commented:
> in #36 @KentaTada mentioned that we need to read another file to verify cgroupv2.

It seems that we need to check in different ways for cgroup v1 and cgroup v2.

For instance,
- cgroup v2: check `/sys/fs/cgroup/init.scope/cgroup.freeze`
- cgroup v1: check CONFIG_CGROUP_FREEZER.

/hold

pacoxu commented:
```
	// /proc/cgroups is meaningless for v2
	// https://github.com/torvalds/linux/blob/v5.3/Documentation/admin-guide/cgroup-v2.rst#deprecated-v1-core-features
	if IsCgroup2UnifiedMode() {
		// "pseudo" controllers do not appear in /sys/fs/cgroup/cgroup.controllers.
		// - devices: implemented in kernel 4.15
		// - freezer: implemented in kernel 5.2
		// We assume these are always available, as it is hard to detect availability.
		pseudo := []string{"devices", "freezer"}
		data, err := ReadFile("/sys/fs/cgroup", "cgroup.controllers")
		if err != nil {
			return nil, err
		}
		subsystems := append(pseudo, strings.Fields(data)...)
		return subsystems, nil
	}
```
Runc code  https://github.com/kubernetes/kubernetes/blob/e404a28fae6946483c5de0e9f4b4adf6e6f63f3b/vendor/github.com/opencontainers/runc/libcontainer/cgroups/utils.go#L100-L114

pacoxu commented:
https://github.com/kubernetes/system-validators/blob/c4cf63549977e8d75add5bb76a1be4acfc727793/validators/cgroup_validator_linux.go#L141-L156

```
	pseudo := []string{"devices", "freezer"}
```
We currently use the similar logic here and add `freezer` as default.

https://github.com/kubernetes/system-validators/blob/c4cf63549977e8d75add5bb76a1be4acfc727793/validators/types_unix.go#L69-L70

This checks the cgroup sub controllers.

neolit123 commented:
aligning with RUNC seems good to me. 

what about these comments about freeze vs freezer:
https://github.com/kubernetes/system-validators/issues/36#issuecomment-2285262696

pacoxu commented:
/unhold
For cgroup v1 cleanup, we should not do it now.

> what about these comments about freeze vs freezer?
> [#36 (comment)](https://github.com/kubernetes/system-validators/issues/36#issuecomment-2285262696)

Updated.

neolit123 reviewed (commented):
/lgtm
/approve
/hold

@KentaTada could you please have a look too?

neolit123 commented:
> For cgroup v1 cleanup, we should not do it now.

could you explain why?

pacoxu commented:
> > For cgroup v1 cleanup, we should not do it now.
> 
> could you explain why?

Basically two reasons:
1. Kubernetes 1.31: Moving cgroup v1 Support into Maintenance Mode, not deprecated.
2. Most OS distrbutions keeps the cgroup v1 related kernel config as is, and I think this may be for backward compatibility.

neolit123 commented:
> > > For cgroup v1 cleanup, we should not do it now.
> > 
> > 
> > could you explain why?
> 
> Basically two reasons IMO:
> 
> 1. Kubernetes 1.31: Moving cgroup v1 Support into Maintenance Mode, not deprecated.
> 2. Most OS distrbutions keeps the cgroup v1 related kernel config as is, and I think this may be for backward compatibility.

ok, got it. you mean that we should not cleanup cgroups v1 because they are still widely used.

i tried finding @KentaTada 's slack to give him a ping but couldn't.
let's wait until end of next week maybe and if he does not reply we can go merge this PR as it is.

KentaTada commented:
> @KentaTada could you please have a look too?

I'll take a look at this pr by the first half of the next week.

KentaTada commented on validators/cgroup_validator_linux.go:
```diff
@@ -142,15 +142,19 @@ func (c *CgroupsValidator) getCgroupV2Subsystems() ([]string, error) {
-	// - freezer: implemented in kernel 5.2
-	pseudo := []string{"devices", "freezer"}
+	subsystems := []string{"devices"}
+	_, err := os.Stat("/sys/fs/cgroup/init.scope/cgroup.freeze")
```
It might be a corner case, but we want to get a path of cgroup.freeze
in a way that is independent of the environment.
- "/sys/fs/cgroup" is the default mountpoint and modifiable.
- "init.scope" is specific for systemd.

pacoxu replied:
> * "/sys/fs/cgroup" is the default mountpoint and modifiable.

https://github.com/kubernetes/system-validators/blob/1d2f5889f07347b9f2d3ed337b3cfb08d45f4cf6/validators/cgroup_validator_linux.go#L48

We currently use this default here for other checks.

If not, we may have to check the `mount`  or `/proc/mounts`/`/proc/self/mounts` to get the cgroup2 mount path. 


> "init.scope" is specific for systemd.

Or we may use `system.slice/cgroup.freeze`?

pacoxu replied:
```
	_, err := os.Stat(filepath.Join(unifiedMountpoint, "system.slice", "/cgroup.freeze"))
```
Changed to this. 

https://github.com/kubernetes/system-validators/blob/4d61a4e8f440a37c88e281ae115e43e8893683a1/validators/cgroup_validator_linux.go#L150-L154

KentaTada replied:
Thank you for your suggestion.

> If not, we may have to check the `mount` or `/proc/mounts`/`/proc/self/mounts` to get the cgroup2 mount path.

Yes. I think we need to get the path from /proc/mounts.
@neolit123  Is this change ok? It may affect the behavior of existing check.

> Or we may use system.slice/cgroup.freeze?

system.slice is also dependent on systemd.
Creating a directory in cgroupfs for this check is the perfect solution.

neolit123 replied:
unfortunately i don't know a lot about the details here.

one point to make - kubeadm supports the openrc init system from alpine and not only systemd:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/initsystem/initsystem_unix.go#L46C14-L46C20

also, k8s kubelet currently only works with the systemd cgroups driver AFAIK.
but for runc it's not mandatory:
https://github.com/opencontainers/runc/blob/main/docs/cgroup-v2.md#systemd 

> system.slice is also dependent on systemd.

yes there are still users using the cgroupfs driver, which means that system.slice will not exist.

> Creating a directory in cgroupfs for this check is the perfect solution.

could you elaborate why we need to create a dir?

IIUC what we need to do to check if a controller foo is enabled is to do the following:
```
check if these paths exist:
  - for cgroupfs /sys/fs/cgroup/freezer
   - for systemd /sys/fs/cgroup/systemd/freezer.state
```
if one of them exists we can say freezer is enabled, i don't think we need to check freezer.state for values.

pacoxu replied:
Updated.

You can see https://github.com/kubernetes/system-validators/compare/4d61a4e8f440a37c88e281ae115e43e8893683a1..2ec937dbfb62d8fa673a16e73db0fa21eb042d3b. 

@KentaTada Do you have time to take another look?

neolit123 replied:
related
- https://github.com/k0sproject/k0s/issues/3069

KentaTada replied:
> could you elaborate why we need to create a dir?

/sys/fs/cgroup/freezer does not exist.
Like k0s, we also need to create a dummy cgroup:
https://github.com/k0sproject/k0s/pull/3077/files#diff-468315026edb0a5cee0863f23231ad03809dcf5aa33958e7df1fd03779aa2621R171

KentaTada replied:
> @KentaTada Do you have time to take another look?

I took another look.
Could you confirm it?

KentaTada commented on validators/types_unix.go:
```diff
@@ -40,11 +40,12 @@ var DefaultSysSpec = SysSpec{
 			{Name: "CGROUP_FREEZER"},
 			{Name: "CGROUP_PIDS"},
-			{Name: "CGROUP_SCHED"},
+			{Name: "CGROUP_SCHED"}, // cgroups v1 cpu
```
It seems that Utilization Clamping depends on CONFIG_CGROUP_SCHED.
https://github.com/torvalds/linux/blob/v6.12-rc1/init/Kconfig#L1080
Since the values of Utilization Clamping can be set through the cgroup
v2 interface, CONFIG_CGROUP_SCHED must also be enabled for cgroup v2.
https://github.com/torvalds/linux/blob/v6.12-rc1/kernel/sched/core.c#L9920

pacoxu replied:
Updated.

KentaTada commented on validators/types_unix.go:
```diff
@@ -40,11 +40,12 @@ var DefaultSysSpec = SysSpec{
+			{Name: "CGROUP_BPF"},     // cgroups v2
+			{Name: "CGROUP_CPUACCT"}, // cgroups v1 cpuacct
 			{Name: "CGROUP_DEVICE"},
 			{Name: "CGROUP_FREEZER"},
```
CONFIG_CGROUP_FREEZER is for cgroups v1 freezer.
According to the below, we may disable CONFIG_CGROUP_FREEZER when we
use cgroups v2 freezer.
https://github.com/torvalds/linux/blob/v6.12-rc1/init/Kconfig#L1132

KentaTada commented:
@pacoxu @neolit123 
I'm sorry for the late reply.
I left some comments.
In addition to that, I think we should split the structure of the required kernel config into two parts: one for v1 and the other for v2?
If users are using v2, they won’t want to see the list for v1.

pacoxu commented:
BTW, I am trying to write a blog about this in https://github.com/kubernetes/website/pull/47342/. Could you help to review as well？

>> In addition to that, I think we should split the structure of the required kernel config into two parts: one for v1 and the other for v2?
>> If users are using v2, they won’t want to see the list for v1.
> IIUC, the only way to split the v1 and v2 spec is to introduce a breaking change in the spec API, which means we need to release a new MAJOR version of this small library.

If so, may we do  that later?

neolit123 commented:
there seem to be code org benefits from it, but overall, as long as we add the proper comment in the spec we can avoid changing the spec API?

i don't know how many users this lib has, but likely there are not many.
either way, we should follow semver rules for it.

pacoxu commented:
I would proposal to do this in comment now for the current release. 
And later, we can refactor and release a new MAJOR version. Do we need a tracking issue or major release branch for that(it seems that we don't have other new features here)?

neolit123 commented:
> I would proposal to do this in comment now for the current release. And later, we can refactor and release a new MAJOR version. Do we need a tracking issue or major release branch for that(it seems that we don't have other new features here)?

i don't think we need a tracking issue.
also, thinking more about it we should schedule this breaking change with another one. i.e. don't release a MAJOR version only because of the cgroups API refactor.

KentaTada commented:
> i don't think we need a tracking issue.
> also, thinking more about it we should schedule this breaking change with another one. i.e. don't release a MAJOR version only because of the cgroups API refactor.

OK. I agree with you.
Because don't need to split the v1 and v2 spec, I think we should just leave a suitable comment.

KentaTada commented:
> BTW, I am trying to write a blog about this in [kubernetes/website#47342](https://github.com/kubernetes/website/pull/47342). Could you help to review as well？

Yes!
I'd like to review it.
Thanks!

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -142,15 +165,39 @@ func (c *CgroupsValidator) getCgroupV2Subsystems() ([]string, error) {
+			fmt.Printf("error remove dir %q: %v\n", tmpDir, err)
+		}
+	}()
+	_, err = os.Stat(filepath.Join(tmpDir, "/cgroup.freeze"))
```
```suggestion
	_, err = os.Stat(filepath.Join(tmpDir, "cgroup.freeze"))
```

is that additional / needed?

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -142,15 +165,39 @@ func (c *CgroupsValidator) getCgroupV2Subsystems() ([]string, error) {
 }
+
+// For freezer which is implemented in kernel 5.2, we can check the existence of `cgroup.freeze`.
+func checkCgroupV2Freeze() bool {
```
```suggestion
func checkCgroupV2Freeze() (error, bool) {
```

instead of print return error and bool

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -142,15 +165,39 @@ func (c *CgroupsValidator) getCgroupV2Subsystems() ([]string, error) {
+	defer func() {
+		err := os.RemoveAll(tmpDir)
+		if err != nil {
+			fmt.Printf("error remove dir %q: %v\n", tmpDir, err)
```
```suggestion
			fmt.Printf("error removing directory %q: %v\n", tmpDir, err)
```

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -142,15 +165,39 @@ func (c *CgroupsValidator) getCgroupV2Subsystems() ([]string, error) {
+func checkCgroupV2Freeze() bool {
+	tmpDir, err := os.MkdirTemp(unifiedMountpoint, "freezer-test")
+	if err != nil {
+		fmt.Printf("error mkdir under %q: %v\n", unifiedMountpoint, err)
```
```suggestion
		fmt.Printf("error crearing directory under %q: %v\n", unifiedMountpoint, err)
```

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -142,15 +165,39 @@ func (c *CgroupsValidator) getCgroupV2Subsystems() ([]string, error) {
 	return subsystems, nil
 }
+
+// For freezer which is implemented in kernel 5.2, we can check the existence of `cgroup.freeze`.
```
// To check if the freezer controller is enabled in Linux kernels 5.2, we can check the existence of `cgroup.freeze`.

neolit123 reviewed (commented):
i don't know if creating the tmp dir is the right solution.
leaving review to @KentaTada

KentaTada commented on validators/cgroup_validator_linux.go:
```diff
@@ -44,10 +47,31 @@ func (c *CgroupsValidator) Name() string {
+// getUnifiedMountpoint will check /proc/mounts and return the cgroup dir
+func getUnifiedMountpoint() string {
+	f, err := os.Open(mountsFilePath)
+	if err != nil {
```
fstype should be checked to confirm whether defaultUnifiedMountpoint is actually cgroupfs.

pacoxu replied:
if we cannot open the `/proc/mounts`(err!=nil here), how can we confirm about the default mountpoint groupfs?

- check `/proc/self/mountinfo`?

pacoxu replied:
https://github.com/kubernetes/system-validators/pull/41/commits/95629f0e852ddc8c74bdb25c98cf144509146018 tried to address this.

 1. check /proc/self/mountinfo if the default mount point is available
 2. check /proc/mounts and return the cgroup dir

pacoxu replied:
@neolit123 @KentaTada  PTAL

KentaTada commented on validators/cgroup_validator_linux.go:
```diff
@@ -44,10 +47,31 @@ func (c *CgroupsValidator) Name() string {
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		// example fields: `cgroup2 /sys/fs/cgroup cgroup2 rw,seclabel,nosuid,nodev,noexec,relatime 0 0`
+		fields := strings.Split(scanner.Text(), " ")
```
Just in case, I think that strings.Fields is suitable to handle spaces like below:
https://github.com/kubernetes/system-validators/blob/1d2f5889f07347b9f2d3ed337b3cfb08d45f4cf6/validators/cgroup_validator_linux.go#L132

pacoxu replied:
Updated.

KentaTada commented on validators/cgroup_validator_linux.go:
```diff
@@ -44,10 +47,31 @@ func (c *CgroupsValidator) Name() string {
+	for scanner.Scan() {
+		// example fields: `cgroup2 /sys/fs/cgroup cgroup2 rw,seclabel,nosuid,nodev,noexec,relatime 0 0`
+		fields := strings.Split(scanner.Text(), " ")
+		if len(fields) >= 3 && (fields[2] == "cgroup2" || fields[2] == "cgroup") {
```
Which shoud we select the mount point if there are multiple cgroup mount points?

pacoxu replied:
The current logic is to return the first `cgroup` or `cgroup2`.

Do you mean that we can return `cgroup2` if we have? And if not, we return the `cgroup` mount point.

pacoxu replied:
Updated.

KentaTada commented on validators/cgroup_validator_linux.go:
```diff
@@ -142,15 +166,41 @@ func (c *CgroupsValidator) getCgroupV2Subsystems() ([]string, error) {
+		}
+	}()
+	_, err = os.Stat(filepath.Join(tmpDir, "cgroup.freeze"))
+	if err == nil {
```
Can we use os.IsNotExist(err) to clarify the root cause?
Not the absence of the cgroup.freeze but other errors may happen.

pacoxu replied:
Updated.
- Should I print error if it is not `NotExistErr`?

pacoxu commented:
I updated the PR accordingly and the only unresolved comment is https://github.com/kubernetes/system-validators/pull/41/files#r1800338762. 
- I am not sure how to confirm about the default mount point if we cannot get the mount info correctly.

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -44,10 +47,96 @@ func (c *CgroupsValidator) Name() string {
 
+// 1. check /proc/self/mountinfo if the default mount point is available
+// 2. check /proc/mounts and return the cgroup dir
+func getUnifiedMountpoint() string {
```
make this function return a string, error instead of calling printf() in it.

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -44,10 +47,96 @@ func (c *CgroupsValidator) Name() string {
 
+// 1. check /proc/self/mountinfo if the default mount point is available
+// 2. check /proc/mounts and return the cgroup dir
+func getUnifiedMountpoint() string {
```
make this function return a string, error instead of calling printf() in it.

pacoxu replied:
Do we need this warning? 
https://github.com/kubernetes/system-validators/pull/41/files#diff-0cb89d807d0cf7f118534018122b93367efce21da4e1650705f6c991872ae3e4R63

neolit123 replied:
i think the url you posted is not correct ^
but basically if this func returns err the caller can decide if it's a warning or error.

pacoxu replied:
https://github.com/kubernetes/system-validators/pull/41/files#r1800594545 This link.

neolit123 replied:
print should only be done from the main Validate() function.
let's try to show this warning on the top caller.

once you update i will check again.

pacoxu replied:
return it as a warning.

Updated.

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -44,10 +47,96 @@ func (c *CgroupsValidator) Name() string {
+	mountInfoPath            = "/proc/self/mountinfo"
 )
 
+// 1. check /proc/self/mountinfo if the default mount point is available
```
```suggestion
// getUnifiedMountpoint does the following:
// 1. check /proc/self/mountinfo if the default mount point is available
```

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -44,10 +47,96 @@ func (c *CgroupsValidator) Name() string {
+		return cgroupv1MountPoint, nil
+	}
+	return "", fmt.Errorf("cannot get a cgroupfs mount point from %q", path)
+
```
```suggestion
```

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -142,15 +231,43 @@ func (c *CgroupsValidator) getCgroupV2Subsystems() ([]string, error) {
+	subsystems := []string{"devices"}
+	freezeSupported, err := checkCgroupV2Freeze()
+	if err != nil {
+		fmt.Printf("error checking cgroups v2 freeze: %v\n", err)
```
instead of printing here i think the function should return []string, error, error
```
		subsystems, warn, err = c.getCgroupV2Subsystems()
```

the caller function has it:
```
func (c *CgroupsValidator) Validate(spec SysSpec) (warns, errs []error) {
```

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -142,15 +231,43 @@ func (c *CgroupsValidator) getCgroupV2Subsystems() ([]string, error) {
 	return subsystems, nil
 }
+
+// To check if the freezer controller is enabled in Linux kernels 5.2, we can check the existence of cgroup.freeze.
+func checkCgroupV2Freeze() (bool, error) {
```
```suggestion
// checkCgroupV2Freeze checks if the freezer controller is enabled in Linux kernels 5.2, we can check the existence of cgroup.freeze.
func checkCgroupV2Freeze() (bool, error) {
```

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -142,15 +231,43 @@ func (c *CgroupsValidator) getCgroupV2Subsystems() ([]string, error) {
+	defer func() {
+		err := os.RemoveAll(tmpDir)
+		if err != nil {
+			fmt.Printf("error removing directory %q: %v\n", tmpDir, err)
```
return warning or error don't print.

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -142,15 +231,43 @@ func (c *CgroupsValidator) getCgroupV2Subsystems() ([]string, error) {
+	if os.IsNotExist(err) {
+		return false, nil
+	} else if err != nil {
+		fmt.Printf("error stat cgroup.freeze file under %q: %v\n", tmpDir, err)
```
same here

pacoxu commented on validators/cgroup_validator_linux.go:
```diff
@@ -44,28 +44,116 @@ func (c *CgroupsValidator) Name() string {
+	if isCgroupfs {
+		return defaultUnifiedMountpoint, err
+	}
+	fmt.Printf("warning: %q is not cgroupfs: %v\n", defaultUnifiedMountpoint, err)
```
This is the only left `Printf`.
Is this warning needed?  Or we just ignore it.

KentaTada commented:
@pacoxu 
Thank you for improving this feature.
It looks good to me.

pacoxu commented:
> @pacoxu
> Thank you for improving this feature.
> It looks good to me.

Thanks for your detailed review and comments. 
Wait for @neolit123 to double check if anything is missing.

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -44,28 +44,117 @@ func (c *CgroupsValidator) Name() string {
+// getUnifiedMountpoint checks if the default mount point is available.
+// If not, it parses the mounts file to find a valid cgroup mount point.
+func getUnifiedMountpoint() (string, error, error) {
+	isCgroupfs, err := isMountPointCgroupfs(defaultUnifiedMountpoint)
+	if isCgroupfs {
+		return defaultUnifiedMountpoint, err, nil
+	}
```
```suggestion
	isCgroupfs, err := isMountPointCgroupfs(defaultUnifiedMountpoint)
	if err != nil {
		return "", err, nil
	}
	if isCgroupfs {
		return defaultUnifiedMountpoint, nil, nil
	}
```
otherwise the err returned by isMountPointCgroupfs is never processed.

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -44,28 +44,117 @@ func (c *CgroupsValidator) Name() string {
+		return defaultUnifiedMountpoint, err, nil
+	}
+	mountPath, err := parseMountsFile(mountsFilePath)
+	return mountPath, err, fmt.Errorf("%q is not cgroupfs: %v", defaultUnifiedMountpoint, err)
```
```suggestion
	return mountPath, nil, warn
```

warn can be constructed earlier.

should it be here?

```
	isCgroupfs, err := isMountPointCgroupfs(defaultUnifiedMountpoint)
	if err != nil {
		warn := fmt.Errorf("%q is not cgroupfs: %v", defaultUnifiedMountpoint, err)
	}
```

i got confused what was the err and what was the warn

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -44,28 +44,117 @@ func (c *CgroupsValidator) Name() string {
+	if isCgroupfs {
+		return defaultUnifiedMountpoint, err, nil
+	}
+	mountPath, err := parseMountsFile(mountsFilePath)
```
```suggestion
	mountPath, err := parseMountsFile(mountsFilePath)
	if err != nil {
		return "", err, warn
	}
```

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -44,28 +44,117 @@ func (c *CgroupsValidator) Name() string {
+	return mountPath, err, fmt.Errorf("%q is not cgroupfs: %v", defaultUnifiedMountpoint, err)
+}
+
+func isMountPointCgroupfs(path string) (bool, error) {
```
i think this should be unit tested.

you can break it down into two functions:
isMountPointCgroupfs opens that file and reads the []byte / string with isMountPointCgroupfsImpl()

impl == implementation. we it this way in some places in kubeadm.
isMountPointCgroupfsImpl should be unit tested.

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -44,28 +44,117 @@ func (c *CgroupsValidator) Name() string {
+		if !strings.Contains(line, "cgroup") {
+			continue
+		}
+		// example fields: `28 22 0:25 / /sys/fs/cgroup rw,nosuid,nodev,noexec,relatime shared:4 - cgroup2 cgroup2 rw,seclabel`
```
the format should be consistent across kernels and distros, is that true?

pacoxu replied:
This is following k/k code and I removed some logics that we did not depends on.

https://github.com/kubernetes/kubernetes/blob/07e73682b9a5259cacac978346ed4567588e4165/staging/src/k8s.io/mount-utils/mount_helper_unix.go#L102-L165

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -44,28 +44,117 @@ func (c *CgroupsValidator) Name() string {
+		}
+		// example fields: `28 22 0:25 / /sys/fs/cgroup rw,nosuid,nodev,noexec,relatime shared:4 - cgroup2 cgroup2 rw,seclabel`
+		fields := strings.Fields(line)
+		if len(fields) >= 9 {
```
what does it mean if there are less fields on the line?

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -44,28 +44,117 @@ func (c *CgroupsValidator) Name() string {
+			if fields[5] == path {
+				i := 6
+				for i < len(fields) && fields[i] != "-" {
+					i++
```
can this iterator be avoided?

pacoxu replied:
The current code is following the k/k. To make is simpler, I think we may split it using ` - `.
- There should be at least 3 fields after `-`.

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -44,28 +44,117 @@ func (c *CgroupsValidator) Name() string {
+					i++
+				}
+				if len(fields)-i <= 3 {
+					return false, fmt.Errorf("expect 3 fields after `-` in mount info file %q", path)
```
```suggestion
					return false, fmt.Errorf("in mount info file %q, expected 3 fields after `-`, but got %d", path, xxxx)
```

also why can this happen?

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -44,28 +44,117 @@ func (c *CgroupsValidator) Name() string {
+	return false, fmt.Errorf("path %q is not cgroupfs", path)
+}
+
+func parseMountsFile(path string) (string, error) {
```
aren't we doing the same thing two times?

- checking a mount point path with isMountPointCgroupfs
- then reading the same file again with this function parseMountsFile

i think this can be done in one go.

here we don't care much about performance but it's just better code org.

pacoxu replied:
```
###  /proc/mounts is the same as  /proc/self/mounts
[root@paco ~]# cat /proc/mounts   | grep cgroup
cgroup2 /sys/fs/cgroup cgroup2 rw,seclabel,nosuid,nodev,noexec,relatime 0 0
none /run/cilium/cgroupv2 cgroup2 rw,seclabel,relatime 0 0
[root@paco ~]# cat /proc/self/mountinfo   | grep cgroup
28 22 0:25 / /sys/fs/cgroup rw,nosuid,nodev,noexec,relatime shared:4 - cgroup2 cgroup2 rw,seclabel
1499 27 0:25 / /run/cilium/cgroupv2 rw,relatime shared:573 - cgroup2 none rw,seclabel
```

Probably, we can just parce the ` /proc/mounts` for the first cgroup v2 path or first cgroup v1 path.

pacoxu commented:
I refactor the logic to just parse `/proc/mounts` to get the first cgroup path.
- and I add a simple test for `getUnifiedMountpoint( )`.

neolit123 commented on validators/cgroup_validator_test.go:
```diff
@@ -93,3 +94,73 @@ func TestValidateCgroupSubsystem(t *testing.T) {
+			path, err := getUnifiedMountpoint(tmpFile.Name())
+
+			if test.expectedErr {
+				assert.Error(t, err, "%q: Expected error but got none", desc)
+			} else {
+				assert.NoError(t, err, "%q: Did not expect error but got one: %s", desc, err)
+			}
+
+			assert.Equal(t, test.expectedPath, path, "%q: Expected cgroup path mismatch", desc)
+		})
```
we can remove the parameter `desc`  when doing assertion e.g.
`assert.Equal(t, test.expectedPath, path, "%q: Expected cgroup path mismatch")`

because the subtest `t.Run(desc, func(t *testing.T) {` already tells which test is currently running.

pacoxu replied:
Good Catch.

neolit123 commented on validators/cgroup_validator_test.go:
```diff
@@ -93,3 +94,73 @@ func TestValidateCgroupSubsystem(t *testing.T) {
+		expectedErr       bool
+		expectedPath      string
+	}{
+		"cgroup v2": {
+			mountsFileContent: "cgroup2 /sys/fs/cgroup cgroup2 rw,seclabel,nosuid,nodev,noexec,relatime 0 0",
+			expectedErr:       false,
+			expectedPath:      "/sys/fs/cgroup",
+		},
+		"cgroup v1": {
```
```suggestion
		"cgroups v2": {
			mountsFileContent: "cgroup2 /sys/fs/cgroup cgroup2 rw,seclabel,nosuid,nodev,noexec,relatime 0 0",
			expectedErr:       false,
			expectedPath:      "/sys/fs/cgroup",
		},
		"cgroups v1": {
```

neolit123 commented on validators/cgroup_validator_test.go:
```diff
@@ -93,3 +94,73 @@ func TestValidateCgroupSubsystem(t *testing.T) {
+			expectedErr:       true,
+			expectedPath:      "",
+		},
+		"none cgroup mounts": {
```
```suggestion
		"no cgroup mounts": {
```

neolit123 commented on validators/cgroup_validator_test.go:
```diff
@@ -93,3 +94,73 @@ func TestValidateCgroupSubsystem(t *testing.T) {
+			expectedErr:  false,
+			expectedPath: "/sys/fs/cgroup/unified",
+		},
+		"cgroup v1 only with multiple subsystems": {
```
```suggestion
		"cgroups v1 only with multiple subsystems": {
```

neolit123 commented on validators/cgroup_validator_test.go:
```diff
@@ -93,3 +94,73 @@ func TestValidateCgroupSubsystem(t *testing.T) {
+			mountsFileContent: `cgroup /sys/fs/cgroup/cpuset cgroup rw,nosuid,nodev,noexec,relatime,cpuset
+cgroup /sys/fs/cgroup/memory cgroup rw,nosuid,nodev,noexec,relatime,memory`,
+			expectedErr:  false,
+			expectedPath: "/sys/fs/cgroup/cpuset", // First valid cgroup v1 path
```
```suggestion
			expectedPath: "/sys/fs/cgroup/cpuset", // First valid cgroups v1 path
```

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -45,27 +45,72 @@ func (c *CgroupsValidator) Name() string {
+		if !strings.Contains(line, "cgroup") {
+			continue
+		}
+		// example fields: `cgroup2 /sys/fs/cgroup cgroup2 rw,seclabel,nosuid,nodev,noexec,relatime 0 0`
```
```suggestion
		// Example fields: `cgroup2 /sys/fs/cgroup cgroup2 rw,seclabel,nosuid,nodev,noexec,relatime 0 0`.
```

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -45,27 +45,72 @@ func (c *CgroupsValidator) Name() string {
+		if len(fields) >= 3 {
+			switch fields[2] {
+			case "cgroup2":
+				// return the first cgroup v2 mount point directly
```
```suggestion
				// Return the first cgroups v2 mount point directly.
```

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -45,27 +45,72 @@ func (c *CgroupsValidator) Name() string {
+				// return the first cgroup v2 mount point directly
+				return fields[1], nil
+			case "cgroup":
+				// set the first cgroup v1 mount point only,
+				// and continue the loop to find if there is a cgroup v2 mount point
```
```suggestion
				// Set the first cgroups v1 mount point only,
				// and continue the loop to find if there is a cgroup v2 mount point.
```

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -45,27 +45,72 @@ func (c *CgroupsValidator) Name() string {
+			}
+		}
+	}
+	// return cgroup v1 mount point if no cgroup v2 mount point is found
```
```suggestion
	// Return cgroups v1 mount point if no cgroups v2 mount point is found.
```

neolit123 commented on validators/cgroup_validator_linux.go:
```diff
@@ -45,27 +45,72 @@ func (c *CgroupsValidator) Name() string {
+			case "cgroup":
+				// set the first cgroup v1 mount point only,
+				// and continue the loop to find if there is a cgroup v2 mount point
+				if len(cgroupV1MountPoint) == 0 {
+					cgroupV1MountPoint = fields[1]
+				}
```
is it ok to just do this?
```suggestion
				cgroupV1MountPoint = fields[1]
```

i.e. always update the var if more cgroups v1 mount points are found or should we only catch the first one?
(i don't know the format)

pacoxu replied:
```
[root@10-29-14-249 ~]# cat /proc/mounts | grep cgroup
tmpfs /sys/fs/cgroup tmpfs ro,seclabel,nosuid,nodev,noexec,mode=755 0 0
cgroup /sys/fs/cgroup/systemd cgroup rw,seclabel,nosuid,nodev,noexec,relatime,xattr,release_agent=/usr/lib/systemd/systemd-cgroups-agent,name=systemd 0 0
cgroup /sys/fs/cgroup/blkio cgroup rw,seclabel,nosuid,nodev,noexec,relatime,blkio 0 0
cgroup /sys/fs/cgroup/hugetlb cgroup rw,seclabel,nosuid,nodev,noexec,relatime,hugetlb 0 0
cgroup /sys/fs/cgroup/perf_event cgroup rw,seclabel,nosuid,nodev,noexec,relatime,perf_event 0 0
cgroup /sys/fs/cgroup/memory cgroup rw,seclabel,nosuid,nodev,noexec,relatime,memory 0 0
cgroup /sys/fs/cgroup/net_cls,net_prio cgroup rw,seclabel,nosuid,nodev,noexec,relatime,net_prio,net_cls 0 0
cgroup /sys/fs/cgroup/cpu,cpuacct cgroup rw,seclabel,nosuid,nodev,noexec,relatime,cpuacct,cpu 0 0
cgroup /sys/fs/cgroup/devices cgroup rw,seclabel,nosuid,nodev,noexec,relatime,devices 0 0
cgroup /sys/fs/cgroup/pids cgroup rw,seclabel,nosuid,nodev,noexec,relatime,pids 0 0
cgroup /sys/fs/cgroup/cpuset cgroup rw,seclabel,nosuid,nodev,noexec,relatime,cpuset 0 0
cgroup /sys/fs/cgroup/freezer cgroup rw,seclabel,nosuid,nodev,noexec,relatime,freezer 0 0
```

For cgroup v1, this is an example. I prefer to get the first one.

neolit123 reviewed (commented):
this is much better @pacoxu !
mostly requesting minor fixes for text / comments in multiple places.

and only one Q about the file parser.

neolit123 reviewed (commented):
/lgtm
/approve

should i create a new release after this?

pacoxu commented:
> should i create a new release after this?

Yes, as we have no todo items  for v1.32.

neolit123 commented:
unrelated fix, please LGTM
https://github.com/kubernetes/system-validators/pull/42

will cut a release after it merges and send PR for k/k.

pacoxu commented:
/unhold
Thanks @KentaTada @neolit123 for your reviewing.

neolit123 commented:
i will also rename the branch of this repo to main
